tdms/src/interpolate_Hfield.cpp

```diff
@@ -1,161 +1,179 @@
-    double *first_interp_vals;
-    double *other_dim_cell_data;
-
-    if (better_scheme(x_scheme, CUBIC_INTERP_MIDDLE)) {
-        // if using a BLi scheme in x,
-        // we need interpolated data for cells i - (scheme+1) through i - (scheme+1) + 7, inclusive
-        i_low = i - (x_scheme+1); i_high = i - (x_scheme+1)+7;
-        // assign memory that we need to store these values
-        first_interp_vals = (double *)malloc(8*sizeof(double));
+    // first, we need to determine the non-x dimension that we will interpolate into, in order to bring the fields parallel to the Yee cell centre
+    // then, we will interpolate in the x-direction
+
+    // determine the schemes available to us in the y- and z-directions
+    const interpScheme &z_scheme = best_interp_scheme(K, k), &y_scheme = best_interp_scheme(J, j);
+    
+    // determine the scheme that we will be using for the x-direction
+    const interpScheme &x_scheme = best_interp_scheme(I, i);
+
+    // now we need to assemble the data to pass to x_scheme
+    // like in the E-field case, we might reserve more memory than we needed if we happen to be using cubic interpolation, but this isn't the end of the world
+    double pass_to_x_scheme[8];
+    // likewise, this array will hold the data for the interpolation in the non-x directions
+    double other_dim_data[8];
+
+    // for each cell between i - (x_scheme.index+1) + x_scheme.first_nonzero_coeff to i - (x_scheme.index) + x_scheme.last_nonzero_coeff, we need to interpolate in the 2nd (coordinate) dimension
+    // we determine this dimension here to avoid an if statement at every value of the loop variable
+    // note that because we are only "moving cells" in the x-direction, whichever of z_scheme and y_scheme is better at one cell will also be the best at every other cell.
+    if (z_scheme.is_better_than(y_scheme)) {
+        // we will be interpolating in the z-direction first, then in x
+        // we need to do this for cells k - (z_scheme.index+1) + z_scheme.first_nonzero_coeff through to k - (z_scheme.index+1) + z_scheme.last_nonzero_coeff
+        for (int ii = x_scheme.first_nonzero_coeff; ii <= x_scheme.last_nonzero_coeff; ii++)
+        {
+            for (int kk = z_scheme.first_nonzero_coeff; kk <= z_scheme.last_nonzero_coeff; kk++)
+            {
+                // gather the data for interpolating in the z dimension through cell i - (x_scheme.index + 1) + ii
+                other_dim_data[kk] = Hxy[k - (z_scheme.index + 1) + kk][j][i - (x_scheme.index + 1) + ii] + Hxz[k - (z_scheme.index + 1) + kk][j][i - (x_scheme.index + 1) + ii];
+            }
+            // interpolate in z to obtain a value in cell i - (x_scheme.index+1) + ii for the Hx field
+            // place this into the appropriate index in the data being passed to the x_scheme
+            pass_to_x_scheme[ii] = z_scheme.interpolate(other_dim_data);
+        }
     }
     else {
```

Comment:
there feels like there could be some more code reuse i.e. 

```c++
other_scheme = z_scheme.is_better_than(y_scheme) ? z_scheme : y_scheme
```

then the loop to set `pass_to_x_scheme`

Reply:
The difficulty is that the `other_dim_data` we need to extract also depends on the scheme we are employing; for example when we use `z_scheme` we need to extract in the `k` index [here](https://github.com/UCL/TDMS/blob/3a0d133c1ace849a3e9bcf926aace00417945ea8/tdms/src/interpolate_Hfield.cpp#L31) and when we use `y_scheme` we need to extract in the `j` index [here](https://github.com/UCL/TDMS/blob/3a0d133c1ace849a3e9bcf926aace00417945ea8/tdms/src/interpolate_Hfield.cpp#L46).

There's probably a clever way around this using pointers, which'll let us get some more code-reuse out of this. To get something going I figured I'd start with two separate loops for when `z` and `y` are better. Suggestions are welcome, pointers in C++ tend to mess with my head at the best of times :sweat_smile:

Reply:
ah I didn't spot that – may be more challenging that I envisaged! 

***

>  pointers in C++ tend to mess with my head at the best of times 😅
you and me both 😅